# Analysis of a laptop store's inventory:

# Reading the Inventory
Use the csv module to read the laptops.csv file and separate the header from the rows.

In [1]:
import csv
with open('laptops.csv') as f:
    reader_file = csv.reader(f)
    rows = list(reader_file)
    header = rows[0]
    rows = rows[1:]
    
    print(header);print();print(rows[:5])

['Id', 'Company', 'Product', 'TypeName', 'Inches', 'ScreenResolution', 'Cpu', 'Ram', 'Memory', 'Gpu', 'OpSys', 'Weight', 'Price']

[['6571244', 'Apple', 'MacBook Pro', 'Ultrabook', '13.3', 'IPS Panel Retina Display 2560x1600', 'Intel Core i5 2.3GHz', '8GB', '128GB SSD', 'Intel Iris Plus Graphics 640', 'macOS', '1.37kg', '1339'], ['7287764', 'Apple', 'Macbook Air', 'Ultrabook', '13.3', '1440x900', 'Intel Core i5 1.8GHz', '8GB', '128GB Flash Storage', 'Intel HD Graphics 6000', 'macOS', '1.34kg', '898'], ['3362737', 'HP', '250 G6', 'Notebook', '15.6', 'Full HD 1920x1080', 'Intel Core i5 7200U 2.5GHz', '8GB', '256GB SSD', 'Intel HD Graphics 620', 'No OS', '1.86kg', '575'], ['9722156', 'Apple', 'MacBook Pro', 'Ultrabook', '15.4', 'IPS Panel Retina Display 2880x1800', 'Intel Core i7 2.7GHz', '16GB', '512GB SSD', 'AMD Radeon Pro 455', 'macOS', '1.83kg', '2537'], ['8550527', 'Apple', 'MacBook Pro', 'Ultrabook', '13.3', 'IPS Panel Retina Display 2560x1600', 'Intel Core i5 3.1GHz', '8GB', '256GB

# Inventory Class
Start implementing a class to represent the inventory. It get the name of the CSV file as argument and reads it into self.header and self.rows.

In [2]:
class Inventory():
    def __init__(self,csv_filename):
        with open(csv_filename) as f:
            read_file = csv.reader(f)
            rows = list(read_file)
        self.header = rows[0];
        self.rows = rows[1:];
        
        for row in self.rows:
            row[-1]=int(row[-1]);
            
            
inventory = Inventory('laptops.csv');

print(inventory.header)
print();
print(len(inventory.rows))
            

['Id', 'Company', 'Product', 'TypeName', 'Inches', 'ScreenResolution', 'Cpu', 'Ram', 'Memory', 'Gpu', 'OpSys', 'Weight', 'Price']

1303


# Finding a Laptop From the Id:
Implement a get_laptop_from_id() function that given a laptop identifier find the row corresponding to that laptop.

In [3]:
class Inventory():
    def __init__(self,csv_filename):
        with open(csv_filename) as f:
            read_file = csv.reader(f)
            rows = list(read_file)
        self.header = rows[0]
        self.rows = rows[1:]
        for row in self.rows:
            row[-1]=int(row[-1]);
    def get_laptop_from_id(self,laptop_id):
        for row in self.rows:
            if row[0]==laptop_id:
                return row
            
        return None
    
inventory = Inventory('laptops.csv');
print(inventory.get_laptop_from_id('3362737'))
print()
print(inventory.get_laptop_from_id('3362736'))
            

['3362737', 'HP', '250 G6', 'Notebook', '15.6', 'Full HD 1920x1080', 'Intel Core i5 7200U 2.5GHz', '8GB', '256GB SSD', 'Intel HD Graphics 620', 'No OS', '1.86kg', 575]

None


# Improving Id Lookups:
Improve the time complexity of finding a laptop with a given id by precomputing a dictionary that maps laptop ids to rows.



In [4]:
class Inventory():
    def __init__(self,csv_filename):
        with open(csv_filename) as f:
            read_file = csv.reader(f);
            rows = list(read_file)
        self.header = rows[0];
        self.rows = rows[1:];
        
        for row in self.rows:
            row[-1]=int(row[-1])
        
        self.id_to_row = {}
        for row  in self.rows:
            self.id_to_row[row[0]]=row
            
    def get_laptop_from_id(self,laptop_id):
        for row in self.rows:
            if row[0]==laptop_id:
                return row
        return None
    
    def get_laptop_from_id_fast(self,laptop_id):
        if laptop_id in self.id_to_row:
            return self.id_to_row[laptop_id]
        return None
    
inventory = Inventory('laptops.csv')

print(inventory.get_laptop_from_id_fast('3362737'));print();
print(inventory.get_laptop_from_id_fast('3362736'))
    

['3362737', 'HP', '250 G6', 'Notebook', '15.6', 'Full HD 1920x1080', 'Intel Core i5 7200U 2.5GHz', '8GB', '256GB SSD', 'Intel HD Graphics 620', 'No OS', '1.86kg', 575]

None


# Comparing Performance:
Compare the performance of both function for id lookup.

In [5]:
import time;
import random;

ids = [str(random.randint(1000000,9999999)) for _ in range(10000)];

inventory = Inventory('laptops.csv');

total_time_no_dict = 0;
for an_id in ids:
    start = time.time()
    inventory.get_laptop_from_id(an_id)
    end = time.time()
    total_time_no_dict += end-start;
    
total_time_dict = 0;
for an_id in ids:
    start = time.time()
    inventory.get_laptop_from_id_fast(an_id)
    end = time.time()
    total_time_dict += end-start;
    
print(total_time_no_dict);print();print(total_time_dict);   

1.4518053531646729

0.006239414215087891


# Analysis
We got:

>1.4518053531646729
>
>0.006239414215087891

We can see a significant improve in performance. If we divide 1.4518 by 0.0062 we see that the new method is about 233 times faster for this input size.

# Two Laptop Promotion
Write a method that finds whether we can spend a given amount of money by purchasing either one or two laptops.

In [6]:
class Inventory():
    def __init__(self,csv_filename):
        with open(csv_filename) as f:
            read_file = csv.reader(f);
            rows = list(read_file);
        self.header = rows[0];
        self.rows = rows[1:];
        
        for row in self.rows:
            row[-1]=int(row[-1]);
            
        self.id_to_row = {};
        for row in self.rows:
            self.id_to_row[row[0]]=row;
    
    def get_laptop_from_id(self,laptop_id):
        for row in self.rows:
            if row[0]==laptop_id:
                return row
        return None
    
    def get_laptop_from_id_fast(self,laptop_id):
        if laptop_id in self.id_to_row:
            return self.id_to_row[laptop_id]
        return None
    
    
    def check_promotion_dollars(self,dollars):
        for row in self.rows:
            if row[-1]==dollars:
                return True
        for row1 in self.rows:
            for row2 in self.rows:
                if row1[-1] + row2[-1]==dollars:
                    return True
        return False


inventory = Inventory('laptops.csv');
print(inventory.check_promotion_dollars(1000));
print();
print(inventory.check_promotion_dollars(442));
   

True

False


# Optimizing Laptop Promotion:
Create a faster version of the promotion method by using the techniques we've learned in the course.



In [7]:
class Inventory():
    def __init__(self,csv_filename):
        with open(csv_filename) as f:
            read_file = csv.reader(f);
            rows = list(read_file);
        self.header = rows[0];
        self.rows = rows[1:];
        
        for row in self.rows:
            row[-1]=int(row[-1]);
        
        self.prices = set();
        for row in self.rows:
            if row[-1] not in self.prices:
                self.prices.add(row[-1])
        
            
        self.id_to_row = {};
        for row in self.rows:
            self.id_to_row[row[0]]=row;
    
    def get_laptop_from_id(self,laptop_id):
        for row in self.rows:
            if row[0]==laptop_id:
                return row
        return None
    
    def get_laptop_from_id_fast(self,laptop_id):
        if laptop_id in self.id_to_row:
            return self.id_to_row[laptop_id]
        return None
    
    
    def check_promotion_dollars(self,dollars):
        for row in self.rows:
            if row[-1]==dollars:
                return True
        for row1 in self.rows:
            for row2 in self.rows:
                if row1[-1] + row2[-1]==dollars:
                    return True
        return False 
    
    def check_promotion_dollars_fast(self,dollars):
        if dollars in self.prices:
            return True
        for row in self.rows:
            if (dollars-row[-1]) in self.prices:
                return True
        return False
    
            
            
inventory = Inventory('laptops.csv');
print(inventory.check_promotion_dollars_fast(1000))
print()
print(inventory.check_promotion_dollars_fast(442))



True

False


# Comparing Promotion Functions
Compare the performance of both methods for the promotion.

In [8]:
prices = [random.randint(100,5000) for _ in range(100)];
inventory = Inventory('laptops.csv')

total_time_no_set = 0;
for price in prices:
    start = time.time();
    inventory.check_promotion_dollars(price)
    end = time.time()
    total_time_no_set += end-start;
    
total_time_set = 0;
for price in prices:
    start = time.time()
    inventory.check_promotion_dollars_fast(price)
    end = time.time()
    total_time_set += end-start;

print(total_time_no_set);print();
print(total_time_set)


3.279162883758545

0.002528667449951172


# Analysis

>3.279162883758545

>0.002528667449951172

We can see a significant improve in performance. If we divide 3.279 by 0.002529 we see that the new method is about 1297 times faster for this input size.

# Finding Laptops Within a Budget
Implement a method for finding the range of indexes of laptops that fall within a budget.

In [9]:
def row_price(row):
    return row[-1]

class Inventory():
    def __init__(self,csv_filename):
        with open(csv_filename) as f:
            read_file = csv.reader(f);
            rows = list(read_file);
        self.header = rows[0];
        self.rows = rows[1:];
        
        for row in self.rows:
            row[-1]=int(row[-1]);
            
            
        self.prices = set();
        for row in self.rows:
            if row[-1] not in self.prices:
                self.prices.add(row[-1])
            
        self.id_to_row = {};
        for row in self.rows:
            self.id_to_row[row[0]]=row;
        
        
        self.rows_by_price = sorted(self.rows,key=row_price)
        
            

    def get_laptop_from_id(self,laptop_id):
        for row in self.rows:
            if row[0]==laptop_id:
                return row
        return None
    
    def get_laptop_from_id_fast(self,laptop_id):
        if laptop_id in self.id_to_row:
            return self.id_to_row[laptop_id]
        return None
    
    
    def check_promotion_dollars(self,dollars):
        for row in self.rows:
            if row[-1]==dollars:
                return True
        for row1 in self.rows:
            for row2 in self.rows:
                if row1[-1] + row2[-1]==dollars:
                    return True
        return False
    
    def check_promotion_dollars_fast(self,dollars):
        if dollars in self.prices:
            return True
        for row in self.rows:
            if (dollars-row[-1]) in self.prices:
                return True
        return False
    
    def find_first_laptop_more_expensive(self,target_price):
        index_start = 0;
        index_end = len(self.rows_by_price)-1;
        
        #print(index_start,index_end)
        
        while index_start < index_end:
            index_middle = (index_start+index_end)//2;
            price = self.rows_by_price[index_middle][-1];
            if price==target_price:
                return index_middle + 1
            elif price < target_price:
                index_start = index_middle + 1;
            else:
                index_end = index_middle;
        
        price = self.rows_by_price[index_start][-1]
        if price<=target_price:
            return -1
        
        return index_start;
    
inventory = Inventory('laptops.csv')  

print(inventory.find_first_laptop_more_expensive(1000))  
print()
print(inventory.find_first_laptop_more_expensive(10000)) 

683

-1


# Finding Laptops Within a Price Range
Implement a method for finding the range of indexes of laptops that fall within a price range.

In [10]:
def row_price(row):
    return row[-1]

class Inventory():
    def __init__(self,csv_filename):
        with open(csv_filename) as f:
            read_file = csv.reader(f);
            rows = list(read_file);
        self.header = rows[0];
        self.rows = rows[1:];
        
        for row in self.rows:
            row[-1]=int(row[-1]);
            
            
        self.prices = set();
        for row in self.rows:
            if row[-1] not in self.prices:
                self.prices.add(row[-1])
            
        self.id_to_row = {};
        for row in self.rows:
            self.id_to_row[row[0]]=row;
        
        
        self.rows_by_price = sorted(self.rows,key=row_price)
        
            

    def get_laptop_from_id(self,laptop_id):
        for row in self.rows:
            if row[0]==laptop_id:
                return row
        return None
    
    def get_laptop_from_id_fast(self,laptop_id):
        if laptop_id in self.id_to_row:
            return self.id_to_row[laptop_id]
        return None
    
    
    def check_promotion_dollars(self,dollars):
        for row in self.rows:
            if row[-1]==dollars:
                return True
        for row1 in self.rows:
            for row2 in self.rows:
                if row1[-1] + row2[-1]==dollars:
                    return True
        return False
    
    def check_promotion_dollars_fast(self,dollars):
        if dollars in self.prices:
            return True
        for row in self.rows:
            if (dollars-row[-1]) in self.prices:
                return True
        return False
    
    def find_first_laptop_more_expensive(self,target_price):
        index_start = 0;
        index_end = len(self.rows_by_price)-1;
        
        
        while index_start < index_end:
            index_middle = (index_start+index_end)//2;
            price = self.rows_by_price[index_middle][-1];
            if price==target_price:
                return index_middle + 1
            elif price < target_price:
                index_start = index_middle + 1;
            else:
                index_end = index_middle;
        
        price = self.rows_by_price[index_start][-1]
        if price<=target_price:
            return -1
        
        return index_start;
    
    def find_first_laptop_less_expensive(self,target_price):
        index_start = 0;
        index_end = len(self.rows_by_price)-1;
        
        #print(index_start,index_end)
        
        while index_start < index_end:
            index_middle = (index_start+index_end)//2;
            price = self.rows_by_price[index_middle][-1];
            #print(price,index_start,index_end)
            
            if price==target_price:
                return index_middle - 1
            elif price < target_price:
                index_start = index_middle + 1;
            else:
                index_end = index_middle;
        
        price = self.rows_by_price[index_start-1][-1]
        #print(index_start)
        if price>=target_price:
            return -1
        
        return index_start-1;
    
    
    def find_laptops_within_a_range(self,price_range):
        min_price=price_range[0];
        max_price=price_range[-1];
        
        low_index_min_price = self.find_first_laptop_less_expensive(min_price);
        high_index_min_price=self.find_first_laptop_more_expensive(min_price);
        avg_index_min_price =(low_index_min_price+high_index_min_price)//2;
        
        
        low_index_max_price = self.find_first_laptop_less_expensive(max_price);
        high_index_max_price=self.find_first_laptop_more_expensive(max_price);
        avg_index_max_price =(low_index_max_price+high_index_max_price)//2;
        
    
        low_price = self.rows_by_price[avg_index_min_price][-1]
        high_price = self.rows_by_price[avg_index_max_price][-1];
        
        if low_price<min_price:
            low_price = self.rows_by_price[high_index_min_price][-1]
        if high_price>max_price:
            high_price = self.rows_by_price[low_index_max_price][-1]
            
        print(low_price,high_price)
        print()
        
        for i in range(len(self.rows_by_price)):
            row = self.rows_by_price[i]
            if row[-1]==low_price:
                index_low_price = i ;
            if row[-1]==high_price:
                index_high_price = i;
        
        
        return self.rows_by_price[index_low_price:index_high_price+1]
    
    

    
inventory = Inventory('laptops.csv')  
a_range = [300,500]

print()
res = inventory.find_laptops_within_a_range(a_range)
print(res[0])
print()
print(res[-1])




304 500

['5415577', 'Asus', 'X541NA-PD1003Y (N4200/4GB/500GB/W10)', 'Notebook', '15.6', '1366x768', 'Intel Pentium Quad Core N4200 1.1GHz', '4GB', '500GB HDD', 'Intel HD Graphics 500', 'Windows 10', '2kg', 304]

['3159162', 'HP', '250 G5', 'Notebook', '15.6', '1366x768', 'Intel Pentium Quad Core N3710 1.6GHz', '4GB', '1TB HDD', 'Intel HD Graphics 405', 'Windows 10', '1.96kg', 500]


# Finding the Most Inexpensive Laptop:
Implement a method for finding the most inexpensive laptop for a fixed set of features

In [11]:
def row_price(row):
    return row[-1]

class Inventory():
    def __init__(self,csv_filename):
        with open(csv_filename) as f:
            read_file = csv.reader(f);
            rows = list(read_file);
        self.header = rows[0];
        self.rows = rows[1:];
        
        for row in self.rows:
            row[-1]=int(row[-1]);
            
            
        self.prices = set();
        for row in self.rows:
            if row[-1] not in self.prices:
                self.prices.add(row[-1])
            
        self.id_to_row = {};
        for row in self.rows:
            self.id_to_row[row[0]]=row;
        
        
        self.rows_by_price = sorted(self.rows,key=row_price)
        
            

    def get_laptop_from_id(self,laptop_id):
        for row in self.rows:
            if row[0]==laptop_id:
                return row
        return None
    
    def get_laptop_from_id_fast(self,laptop_id):
        if laptop_id in self.id_to_row:
            return self.id_to_row[laptop_id]
        return None
    
    
    def check_promotion_dollars(self,dollars):
        for row in self.rows:
            if row[-1]==dollars:
                return True
        for row1 in self.rows:
            for row2 in self.rows:
                if row1[-1] + row2[-1]==dollars:
                    return True
        return False
    
    def check_promotion_dollars_fast(self,dollars):
        if dollars in self.prices:
            return True
        for row in self.rows:
            if (dollars-row[-1]) in self.prices:
                return True
        return False
    
    def find_first_laptop_more_expensive(self,target_price):
        index_start = 0;
        index_end = len(self.rows_by_price)-1;
        
        
        while index_start < index_end:
            index_middle = (index_start+index_end)//2;
            price = self.rows_by_price[index_middle][-1];
            if price==target_price:
                return index_middle + 1
            elif price < target_price:
                index_start = index_middle + 1;
            else:
                index_end = index_middle;
        
        price = self.rows_by_price[index_start][-1]
        if price<=target_price:
            return -1
        
        return index_start;
    
    def find_first_laptop_less_expensive(self,target_price):
        index_start = 0;
        index_end = len(self.rows_by_price)-1;
        
        #print(index_start,index_end)
        
        while index_start < index_end:
            index_middle = (index_start+index_end)//2;
            price = self.rows_by_price[index_middle][-1];
            #print(price,index_start,index_end)
            
            if price==target_price:
                return index_middle - 1
            elif price < target_price:
                index_start = index_middle + 1;
            else:
                index_end = index_middle;
        
        price = self.rows_by_price[index_start-1][-1]
        #print(index_start)
        if price>=target_price:
            return -1
        
        return index_start-1;
    
    
    def find_laptops_within_a_range(self,price_range):
        min_price=price_range[0];
        max_price=price_range[-1];
        
        low_index_min_price = self.find_first_laptop_less_expensive(min_price);
        high_index_min_price=self.find_first_laptop_more_expensive(min_price);
        avg_index_min_price =(low_index_min_price+high_index_min_price)//2;
        
        
        low_index_max_price = self.find_first_laptop_less_expensive(max_price);
        high_index_max_price=self.find_first_laptop_more_expensive(max_price);
        avg_index_max_price =(low_index_max_price+high_index_max_price)//2;
        
    
        low_price = self.rows_by_price[avg_index_min_price][-1]
        high_price = self.rows_by_price[avg_index_max_price][-1];
        
        if low_price<min_price:
            low_price = self.rows_by_price[high_index_min_price][-1]
        if high_price>max_price:
            high_price = self.rows_by_price[low_index_max_price][-1]
            
        print(low_price,high_price)
        print()
        
        for i in range(len(self.rows_by_price)):
            row = self.rows_by_price[i]
            if row[-1]==low_price:
                index_low_price = i;
            if row[-1]==high_price:
                index_high_price = i;
        
        
        return self.rows_by_price[index_low_price:index_high_price+1]
    
    
    def find_least_expensive_laptop(self,specs):
        #indices 7,8
        ram = str(specs[0]).upper();
        hd = str(specs[-1]).upper();
        
        row_inexp_lap = [];
        
        for row in self.rows_by_price:
            if ram in row[7]:
                if hd in row[8]:
                    row_inexp_lap.append(row)
             
        return row_inexp_lap
    

#specs = ['8gb','256gb']
specs = [16, 256]
inventory = Inventory("laptops.csv");
res = inventory.find_least_expensive_laptop(specs)
row_inexp_lap_prices = [];
for row in res:
    row_inexp_lap_prices.append(row[-1]);
    
print(len(res));print()
n=3
for i in range(len(res[:n])):
    print(res[i])

most_inexp_price_row = res[0]
most_inexp_price = min(row_inexp_lap_prices);
most_exp_price = max(row_inexp_lap_prices)
most_exp_price_row = res[-1]
print()
print(most_inexp_price_row);print();
print('most_inexp_price: $'+str(most_inexp_price));print()
print(most_exp_price_row);print();
print('most_exp_price: $'+str(most_exp_price));print()


90

['7106466', 'Dell', 'Inspiron 5567', 'Notebook', '15.6', 'Full HD 1920x1080', 'Intel Core i7 7500U 2.7GHz', '16GB', '256GB SSD', 'AMD Radeon R7 M445', 'Linux', '2.36kg', 899]
['9952432', 'Dell', 'Inspiron 5570', 'Notebook', '15.6', 'Full HD 1920x1080', 'Intel Core i7 8550U 1.8GHz', '16GB', '256GB SSD +  2TB HDD', 'AMD Radeon 530', 'Linux', '2.02kg', 1049]
['9732542', 'Lenovo', 'Legion Y520-15IKBN', 'Gaming', '15.6', 'IPS Panel Full HD 1920x1080', 'Intel Core i7 7700HQ 2.8GHz', '16GB', '256GB SSD', 'Nvidia GeForce GTX 1050 Ti', 'Windows 10', '2.5kg', 1079]

['7106466', 'Dell', 'Inspiron 5567', 'Notebook', '15.6', 'Full HD 1920x1080', 'Intel Core i7 7500U 2.7GHz', '16GB', '256GB SSD', 'AMD Radeon R7 M445', 'Linux', '2.36kg', 899]

most_inexp_price: $899

['3764268', 'HP', 'Zbook 17', 'Workstation', '17.3', 'IPS Panel Full HD 1920x1080', 'Intel Xeon E3-1535M v5 2.9GHz', '16GB', '256GB SSD', 'Nvidia Quadro M2000M', 'Windows 7', '3kg', 4389]

most_exp_price: $4389

